File: docs/app/GUI.js
import { COLOR,  MOUSE } from "./Constants.js";
import { Button } from "./Button.js";
import { Rect } from "./Rectangle.js";
import { DrawRect } from "./SupportFunctions.js";
import { Point } from "./Point.js";
import { Edge } from "./Edge.js";
import { Text } from "./Text.js";

const VERTEX = 0;
const EDGE = 1;

const ADD = 0;
const DELETE = 1;

const CLASS_HIGH = 3;
const CLASS_HIGH_MED = 2;
const CLASS_MED_LOW = 1;
const CLASS_LOW = 0;


export class GUI {
    constructor(application_core){
        this.application_core = application_core;
        this.items = [];
        this.points = [];
        this.edges = [];
        this.Initialize();
    }

    Initialize() {


        this.background_color = COLOR.LIGHTGRAY;
        this.background_color_index = 7;
        this.canvas = this.application_core.canvas;
        this.screen = this.canvas.getContext("2d");
        this.canvasRect = this.canvas.getBoundingClientRect();
        
        this.vertexButton = new Button(200, 20, 150, 50, "Vertex", 10);
        this.vertexButton.OnClick = this.SelectVertexOrEdge.bind(this, this.edgeButton);
        this.Add(this.vertexButton);

        this.edgeButton = new Button(380, 20, 150, 50, "Edge", 20);
        this.edgeButton.OnClick = this.SelectVertexOrEdge.bind(this, this.edgeButton);
        this.Add(this.edgeButton);

        this.addButton = new Button(20, 100, 150, 50, "Add", 35);
        this.addButton.OnClick = this.SelectMode.bind(this, this.addButton);
        this.Add(this.addButton);

        this.deleteButton = new Button(20, 170, 150, 50, "Delete", 10);
        this.deleteButton.OnClick = this.SelectMode.bind(this, this.deleteButton);
        this.Add(this.deleteButton);

        this.demoButton = new Button(560, 20, 150, 50, "Demo", 17);
        this.demoButton.OnClick = this.LoadDemo.bind(this);
        this.Add(this.demoButton);

        this.resetButton = new Button(740, 20, 150, 50, "Reset", 20);
        this.resetButton.OnClick = this.ClearCanvas.bind(this);
        this.Add(this.resetButton);

        
        this.lHighLabelText = new Text(20, 250, 20, 20, "L_high:");
        this.Add(this.lHighLabelText);
        this.lHighMedLabelText = new Text(20, 350, 20, 20, "L_high_medium:");
        this.Add(this.lHighMedLabelText);
        this.lMedLowLabelText = new Text(20, 450, 20, 20, "L_medium_low:");
        this.Add(this.lMedLowLabelText);
        this.lLowLabelText = new Text(20, 550, 20, 20, "L_low");
        this.Add(this.lLowLabelText);

        this.lHighText = new Text(20, 275, 20, 20, "[]");
        this.Add(this.lHighText);
        this.lHighMedText = new Text(20, 375, 20, 20, "[]");
        this.Add(this.lHighMedText);
        this.lMedLowText = new Text(20, 475, 20, 20, "[]");
        this.Add(this.lMedLowText);
        this.lLowText = new Text(20, 575, 450, 20, "[]");
        this.Add(this.lLowText);

        this.ClearCanvas();

    }

    DrawGUI(){
        DrawRect(this.screen, this.background_color, 0, 0, this.canvas.width, this.canvas.height, 0);
        DrawRect(this.screen, COLOR.WHITE, 200, 100, this.canvas.width, this.canvas.height, 0);
        
        for (const element of this.items) {
            element.DrawObject(this.screen);
        }

        for (const edge of this.edges) {
            edge.DrawObject(this.screen);
        }
        
        for (const point of this.points) {
            point.DrawObject(this.screen);
        }
        
    }

    Add(item){
        this.items.push(item);
    }

    AddPoint(point) {
        this.points.push(point);
    }

    AddEdge(edge) {
        this.edges.push(edge);
    }

    MouseHandler(event) {
        event["realpos"] = this.EventPos(event);
        if(event.type == "mousemove") {
            if (this.mouseDown) {
                this.OnDrag(event);
            }
        } else if(event.type == "mousedown" || event.type == "wheel") {
            console.log("Qwe");
            
            this.OnClick(event);
        } else { // MouseUp, MouseOut
            this.OnRelease(event);
        }
        
    }

    KeyboardHandler(event) {
        if(event.type == "keyup") {
            if(event.key == "a") {
                this.modeAD = ADD;
            } else if(event.key == "d") {
                this.modeAD = DELETE;
            } else if(event.key == "w") {
                this.modeVE = VERTEX;
            } else if(event.key == "e") {
                this.modeVE = EDGE;
            }
        }
        this.UpdateSelection();
    }

    OnDrag(event){ 
        // PASS
    }

    OnRelease(event) {
        this.mouseDown = false;
        if(event.realpos[0] > 220 && 
            event.realpos[1] > 120 && 
            event.realpos[0] < this.canvas.width - 30 && 
            event.realpos[1] < this.canvas.height - 30 ) {

            if(this.modeVE == VERTEX) {
                if(this.modeAD == ADD) { // VERTEX + ADD
                    if(!this.CoverOtherPoint(event.realpos)){
                        this.CreatePoint(...event.realpos);
                    }
                } else {  // VERTEX + DELETE
                    var marked = null;
                    for (const point of this.points) {
                        if(point.IsInside(event.realpos)) {
                            var edges = point.DisconnectAllEdges();
                            for (const edge of edges) {
                                this.edges.remove(edge);
                            }
                            marked = point;
                        }
                    }
                    if(marked) {
                        this.points.remove(marked);
                        this.ReclassifyAllPoints();
                    }
                }
            } else { 
                if(this.modeAD == ADD) {  // EDGE + ADD
                    for (const point of this.points) {
                        if(point.IsInside(event.realpos)) {
                            if(!this.point1) {
                                point.SelectThis();
                                this.point1 = point;
                                break;
                            } else {
                                if(this.point1 == point) {
                                    point.DeselectThis();
                                    this.point1 = null;
                                } else {
                                    this.point2 = point;
                                    this.point2.SelectThis();
                                }
                                break;
                            }
                        }
                    }
                    if(this.point1 && this.point2) {
                        this.CreateEdge(this.point1, this.point2);
                        this.point1.DeselectThis();
                        this.point2.DeselectThis();
                        this.point1 = null;
                        this.point2 = null;
                    }
                } else {  // EDGE + DELETE
                    for (const point of this.points) {
                        if(point.IsInside(event.realpos)) {
                            if(!this.point1) {
                                point.SelectThis();
                                this.point1 = point;
                                break;
                            } else {
                                if(this.point1 == point) {
                                    point.DeselectThis();
                                    this.point1 = null;
                                } else {
                                    this.point2 = point;
                                    this.point2.SelectThis();
                                }
                                break;
                            }
                        }
                    }
                    if(this.point1 && this.point2) {
                        var edge = this.point1.DisconnectEdge(this.point2);
                        if(edge) {
                            this.edges.remove(edge);
                            this.ReclassifyAllPoints();
                        }
                        this.point1.DeselectThis();
                        this.point2.DeselectThis();
                        this.point1 = null;
                        this.point2 = null;
                    }
                }
            }

            
        }
    }

    OnClick(event) {
        this.mouseDown = true;
        for (const item of this.items) {
            if(item.IsInside(event.realpos)) {
                if (item instanceof Button) {
                    item.OnClick(event);
                    return;
                }
            } 
        }
    }

    EventPos(event) {
        return [Math.round((event.clientX - this.canvasRect.left) / (this.canvasRect.right - this.canvasRect.left) * this.canvas.width),
                    Math.round((event.clientY - this.canvasRect.top) / (this.canvasRect.bottom - this.canvasRect.top) * this.canvas.height)];
    }


    SelectVertexOrEdge(event) {
        if(event == this.edgeButton) {
            this.modeVE = EDGE;
        } else {
            this.modeVE = VERTEX;
        }
        this.UpdateSelection();
    }

    SelectMode(event) {
        if(event == this.addButton) {
            this.modeAD = ADD;
        } else {
            this.modeAD = DELETE;
        }
        this.UpdateSelection();
    }

    UpdateSelection() {

        if(this.modeVE == EDGE) {
            this.edgeButton.SelectThis();
            this.vertexButton.DeselectThis();
        } else {
            this.vertexButton.SelectThis();
            this.edgeButton.DeselectThis();
        }

        if(this.modeAD == ADD) {
            this.addButton.SelectThis();
            this.deleteButton.DeselectThis();
        } else {
            this.addButton.DeselectThis();
            this.deleteButton.SelectThis();
        }

        if(this.point1) {
            this.point1.DeselectThis();
            this.point1 = null;
        }
        if(this.point2) {
            this.point2.DeselectThis();
            this.point2 = null;
        }
    }

    CoverOtherPoint(position) {
        var top_left = [position[0] - 15, position[1] - 20];
        var bottom_left = [position[0] - 15, position[1] + 30];

        var top_right = [position[0] + 30, position[1] - 20];
        var bottom_right = [position[0] + 30, position[1] + 30];

        for (const point of this.points) {
            if(point.IsInside(top_left) ||
                point.IsInside(bottom_left) ||
                point.IsInside(top_right) ||
                point.IsInside(bottom_right)) {
                    return true;
                }
        }

        return false;
    }

    ClearCanvas() {
        this.points = [];
        this.edges = [];
        this.modeAD = ADD;
        this.modeVE = VERTEX;
        this.pointCounter = 0;

        this.highArr=  [];
        this.highMedArr= [];
        this.medLowArr= [];
        this.lowArr = [];

        this.UpdateClassArrayTexts();

        this.lHighText.SetText("[]");
        this.UpdateSelection();
    }

    CreatePoint(x, y) {
        var point = new Point(x, y, this.pointCounter);
        this.AddPoint(point);
        this.pointCounter += 1;
        return point;
    }

    CreateEdge(p1, p2) {
        var edge = new Edge(p1, p2);
        if (p1.AddEdge(edge)) {
            p2.AddEdge(edge);
            this.AddEdge(edge);
        }
        
        this.ReclassifyAllPoints();
    }

    UpdateClassArrayTexts() {
        this.lHighText.SetText("["+ this.highArr.join(", ") + "]");
        this.lHighMedText.SetText("["+ this.highMedArr.join(", ") + "]");
        this.lMedLowText.SetText("["+ this.medLowArr.join(", ") + "]");
        this.lLowText.SetText("["+ this.lowArr.join(", ") + "]");
    }

    ClassifyPoint(point) {
        var m = this.edges.length;
        var degV = point.GetDegree();

        var highValue = m ** 0.75;
        var medValue = m ** 0.5;
        var lowValue = m ** 0.25; 

        if(degV >= highValue) {
            point.SetClassification(CLASS_HIGH);
        } else if ( highValue > degV && degV >= medValue) {
            point.SetClassification(CLASS_HIGH_MED);
        } else if ( medValue > degV && degV >= lowValue ) {
            point.SetClassification(CLASS_MED_LOW);
        } else {
            point.SetClassification(CLASS_LOW);
        }
    }

    ReclassifyAllPoints() {
        for (const point of this.points) {
            this.ClassifyPoint(point);
        }
        this.RecollectClassArrays();
    }

    RecollectClassArrays() {
        this.highArr = [];
        this.highMedArr = [];
        this.medLowArr = [];
        this.lowArr = [];
        for (const point of this.points) {
            if(point.classification == CLASS_HIGH) {
                this.highArr.push(point.text);
            } else if (point.classification == CLASS_HIGH_MED) {
                this.highMedArr.push(point.text);
            } else if (point.classification == CLASS_MED_LOW) {
                this.medLowArr.push(point.text);
            } else {
                this.lowArr.push(point.text);
            }
        }
        this.UpdateClassArrayTexts();
    }

    LoadDemo() {
        this.ClearCanvas();
        // Petersen Graf lesz itt

        // Kulso 5szog
        var p1 = this.CreatePoint(700, 140);
        var p2 = this.CreatePoint(980, 345);
        var p3 = this.CreatePoint(870,675);
        var p4 = this.CreatePoint(525, 675);
        var p5 = this.CreatePoint(420, 340);


        // Belso 5szog
        var p6 = this.CreatePoint(700, 260);
        var p7 = this.CreatePoint(875, 385);
        var p8 = this.CreatePoint(810, 590);
        var p9 = this.CreatePoint(595, 590);
        var p10 = this.CreatePoint(525, 390);

        this.CreateEdge(p1, p2);
        this.CreateEdge(p2, p3);
        this.CreateEdge(p3, p4);
        this.CreateEdge(p4, p5);
        this.CreateEdge(p5, p1);

        this.CreateEdge(p6, p8);
        this.CreateEdge(p7, p9);
        this.CreateEdge(p8, p10);
        this.CreateEdge(p9, p6);
        this.CreateEdge(p10, p7);

        this.CreateEdge(p1, p6);
        this.CreateEdge(p2, p7);
        this.CreateEdge(p3, p8);
        this.CreateEdge(p4, p9);
        this.CreateEdge(p5, p10);

        p1.SetPartOfMis(true);
        p4.SetPartOfMis(true);
        p7.SetPartOfMis(true);
        p8.SetPartOfMis(true);

        
    }
}

File: mis-python/App/GUI.py
# -*- coding: utf-8 -*-
from typing import List

from pygame import MOUSEBUTTONDOWN, MOUSEBUTTONUP, MOUSEMOTION
import pygame
import time

from App.Edge import Edge
from App.Point import Point
from App.Text import Text
from App.Button import Button
from App.Rectangle import Rect
from App.Constants import COLOR as Color, MOUSE, FONT_AWESOME as FA

import os.path, os

VERTEX = 0
EDGE = 1

ADD = 0
DELETE = 1
MIS = 2

CLASS_HIGH = 3
CLASS_HIGH_MED = 2
CLASS_MED_LOW = 1
CLASS_LOW = 0


class GUI:

    def __init__(self, parent):
        self.parent = parent
        self.items = []
        self.points = []
        self.edges = []
        self.initilize()

    def initilize(self):
        self.background_color = Color.LIGHTGRAY
        self.background_color_index = 7

        self.vertexButton = Button(200, 20, 150, 50, "Vertex", 10)
        self.vertexButton.bind(self.SelectVertexOrEdge)
        self.Add(self.vertexButton)

        self.edgeButton = Button(380, 20, 150, 50, "Edge", 20)
        self.edgeButton.bind(self.SelectVertexOrEdge)
        self.Add(self.edgeButton)

        self.addButton = Button(20, 100, 150, 50, "Add", 35)
        self.addButton.bind(self.SelectMode)
        self.Add(self.addButton)

        self.deleteButton = Button(20, 170, 150, 50, "Delete", 10)
        self.deleteButton.bind(self.SelectMode)
        self.Add(self.deleteButton)

        self.misButton = Button(20, 20, 150, 50, "MIS", 35)
        self.misButton.bind(self.SelectMode)
        self.Add(self.misButton)

        self.demoButton = Button(560, 20, 150, 50, "Demo", 17)
        self.demoButton.bind(self.LoadDemo)
        self.Add(self.demoButton)

        self.resetButton = Button(740, 20, 150, 50, "Reset", 20)
        self.resetButton.bind(self.ClearCanvas)
        self.Add(self.resetButton)

        self.lHighLabelText = Text(20, 250, 20, 20, "L_high:")
        self.Add(self.lHighLabelText)

        self.lHighMedLabelText = Text(20, 350, 20, 20, "L_high_mediumn:")
        self.Add(self.lHighMedLabelText)

        self.lMedLowLabelText = Text(20, 450, 20, 20, "L_medium_low:")
        self.Add(self.lMedLowLabelText)

        self.lLowLabelText = Text(20, 550, 20, 20, "L_low:")
        self.Add(self.lLowLabelText)

        self.lHighText = Text(20, 275, 20, 20, "[]")
        self.Add(self.lHighText)
        self.lHighMedText = Text(20, 375, 20, 20, "[]")
        self.Add(self.lHighMedText)
        self.lMedLowText = Text(20, 475, 20, 20, "[]")
        self.Add(self.lMedLowText)
        self.lLowText = Text(20, 575, 20, 20, "[]")
        self.Add(self.lLowText)


        self.ClearCanvas()


    def DrawGUI(self, screen):
        w, h = pygame.display.get_surface().get_size()
        pygame.draw.rect(screen, self.background_color, (0, 0, w, h), 0)
        pygame.draw.rect(screen, Color.WHITE, (200, 100, w, h), 0)

        for element in self.items:
            if self.modeVE == EDGE and element == self.misButton:
                continue
            element.DrawObject(screen)

        for edge in self.edges:
            edge.DrawObject(screen)

        for point in self.points:
            point.DrawObject(screen)

    def Add(self, item):
        self.items.append(item)

    def AddPoint(self, point):
        self.points.append(point)

    def AddEdge(self, edge):
        self.edges.append(edge)


    def MouseHandler(self, event):
        if event.type == MOUSEMOTION:
            pass
        elif event.type == MOUSEBUTTONUP:
            self.OnRelease(event)
        elif event.type == MOUSEBUTTONDOWN:
            self.OnClick(event)

    def KeyboardHandler(self, event):
        if event.key == pygame.K_ESCAPE:
            self.Exit()
        elif event.key == pygame.K_a:
            self.modeAD = ADD
        elif event.key == pygame.K_d:
            self.modeAD = DELETE
        elif event.key == pygame.K_w:
            self.modeVE = VERTEX
        elif event.key == pygame.K_e:
            self.modeVE = EDGE
        elif event.key == pygame.K_m:
            self.modeAD = MIS

        self.UpdateSelection()


    def OnRelease(self, event):
        w, h = pygame.display.get_surface().get_size()
        self.mouseDown = False
        if 220 < event.pos[0] < w - 30 and 120 < event.pos[1] < h -30:
            if self.modeVE == VERTEX:
                if self.modeAD == ADD:
                    self.AddVertex(event)
                elif self.modeAD == MIS:
                    self.ChangeMis(event)
                else:
                    self.DeleteVertex(event)
            else:
                if self.modeAD == ADD:
                    self.AddEdgeToGraph(event)
                else:
                    self.DeleteEdgeFromGraph(event)

    def AddVertex(self, event):
        if not self.CoverOtherPoint(event.pos):
            self.CreatePoint(*event.pos)

    def DeleteVertex(self, event):
        marked = None
        for point in self.points:
            if point.IsInside(event.pos):
                otherPoint = None
                edges = point.DisconnectAllEdges()
                for edge in edges:
                    if edge.point1 == point:
                        self.ClassifyPoint(edge.point2)
                        otherPoint = edge.point2
                    else:
                        self.ClassifyPoint(edge.point1)
                        otherPoint = edge.point1

                    self.edges.remove(edge)
                    self.CalculateMISAfterEdgeRemove(point, otherPoint)
                marked = point
        if marked:
            self.points.remove(marked)

    def AddEdgeToGraph(self, event):
        for point in self.points:
            if point.IsInside(event.pos):
                if not self.point1:
                    point.SelectThis()
                    self.point1 = point
                    break
                else:
                    if self.point1 == point:
                        point.DeselectThis()
                        self.point1 = None
                    else:
                        self.point2 = point
                        self.point2.SelectThis()
                    break
        if self.point1 and self.point2:
            self.CreateEdge(self.point1, self.point2)
            self.point1.DeselectThis()
            self.point2.DeselectThis()
            self.point1 = None
            self.point2 = None


    def DeleteEdgeFromGraph(self, event):
        for point in self.points:
            if point.IsInside(event.pos):
                if not self.point1:
                    point.SelectThis()
                    self.point1 = point
                    break
                else:
                    if self.point1 == point:
                        point.DeselectThis()
                        self.point1 = None
                    else:
                        self.point2 = point
                        self.point2.SelectThis()
                    break
        if self.point1 and self.point2:
            edge = self.point1.DisconnectEdge(self.point2)
            if edge:
                self.edges.remove(edge)
                self.ClassifyPoint(self.point1)
                self.ClassifyPoint(self.point2)
                self.CalculateMISAfterEdgeRemove(self.point1, self.point2)
            self.point1.DeselectThis()
            self.point2.DeselectThis()
            self.point1 = None
            self.point2 = None

    def OnClick(self, event):
        self.mouseDown = True
        for item in self.items:
            if item.IsInside(event.pos):
                if isinstance(item, Button):
                    item.OnClick(event, item)
                    return

    def SelectVertexOrEdge(self, event, context):
        if context == self.edgeButton:
            self.modeVE = EDGE
        else:
            self.modeVE = VERTEX

        self.UpdateSelection()

    def SelectMode(self, event, context):
        if context == self.addButton:
            self.modeAD = ADD
        elif context == self.misButton:
            self.modeAD = MIS
        else:
            self.modeAD = DELETE

        self.UpdateSelection()


    def UpdateSelection(self):
        if self.modeVE == EDGE:
            self.edgeButton.SelectThis()
            self.vertexButton.DeselectThis()
            if self.modeAD == MIS:
                self.modeAD = ADD
        else:
            self.vertexButton.SelectThis()
            self.edgeButton.DeselectThis()

        if self.modeAD == ADD:
            self.addButton.SelectThis()
            self.deleteButton.DeselectThis()
            self.misButton.DeselectThis()
        elif self.modeAD == MIS:
            self.addButton.DeselectThis()
            self.deleteButton.DeselectThis()
            self.misButton.SelectThis()
        else:
            self.addButton.DeselectThis()
            self.deleteButton.SelectThis()
            self.misButton.DeselectThis()

        if self.point1:
            self.point1.DeselectThis()
            self.point1 = None

        if self.point2:
            self.point2.DeselectThis()
            self.point2 = None

    def CoverOtherPoint(self, position):
        top_left = (position[0] - 15, position[1] - 20)
        bottom_left = (position[0] - 15, position[1] + 30)

        top_right = (position[0] + 30, position[1] - 20)
        bottom_right = (position[0] + 30, position[1] + 30)

        for point in self.points:
            if point.IsInside(top_left) or \
                    point.IsInside(bottom_left) or \
                    point.IsInside(top_right) or \
                    point.IsInside(bottom_right):
                return True

        return False

    def ClearCanvas(self, event = None, context = None):
        self.maximumIndependentSet = []
        self.points = []
        self.edges = []
        self.modeAD = ADD
        self.modeVE = VERTEX
        self.pointCounter = 0

        self.highArr = []
        self.highMedArr = []
        self.medLowArr = []
        self.lowArr = []

        self.UpdateClassArrayTexts()
        self.lHighText.SetText("[]")
        self.point1 = None
        self.point2 = None
        self.UpdateSelection()

    def CreatePoint(self, x, y):
        point = Point(x, y, self.pointCounter)
        point.SetPartOfMis(True)
        self.AddPoint(point)
        self.pointCounter += 1
        return point

    def CreateEdge(self, p1, p2):
        edge = Edge(p1, p2)
        if p1.AddEdge(edge):
            p2.AddEdge(edge)
            self.AddEdge(edge)

        self.ClassifyPoint(p1)
        self.ClassifyPoint(p2)

        if p1.IsPartOfMis() and p2.IsPartOfMis():
            p1.SetPartOfMis(False)
            p2.SetPartOfMis(True)
            L = list(set(filter(lambda x: x.classification == CLASS_LOW, p1.neighbours)))
            for w in p1.neighbours:
                if w not in L:
                    if w.MIS_neighbours() == 0:
                        w.SetPartOfMis(True)
                else:
                    L_mis = []
                    L_1hop = []

                    a = []

                    for point in w.mis_2hop_neighbours:
                        if point in L:
                            a.append(point)

                    L_2hop = a

                    l_mis = len(L_mis)
                    l_1hop = len(L_1hop)
                    l_2hop = len(L_2hop)

                    if l_2hop <= 4 * len(self.edges) ** 0.75:
                        for point in L_2hop:
                            if point.MIS_neighbours() == 0:
                                L_1hop.append(point)

                        if l_1hop <= 4 * len(self.edges) ** 0.5:
                            for point in L_1hop:
                                if point.real_mis_neighbours == 0:
                                    L_mis.append(point)
                            for point in L_mis:
                                if point.MIS_neighbours() == 0:
                                    point.SetPartOfMis(True)
                        else:
                            for p1 in L_1hop:
                                arr_to_check = p1.mis_neighbours.copy()
                                while len(arr_to_check):
                                    marked = []
                                    for point in arr_to_check:
                                        if point.classification == CLASS_HIGH and \
                                            point.Real_MIS_Neighbours() != 0:
                                            marked.append(point)
                                    for point in marked:
                                        point.SetPartOfMis(False)
                                    arr_to_check = marked.copy()
                    else:
                        for point in L_2hop:
                            if point.real_mis_neighbours == 0:
                                point.SetPartOfMis(True)

                        for point in L_2hop:
                            arr_to_check = point.real_mis_neighbours.copy()
                            while arr_to_check:
                                marked = []
                                for point in arr_to_check:
                                    if point.classification in (CLASS_MED_LOW, CLASS_LOW) and \
                                            point.Real_MIS_Neighbours() != 0:
                                        marked.append(point)
                                for point in marked:
                                    point.SetPartOfMis(False)
                                arr_to_check = marked.copy()

    def CalculateL2Hop(self, L):
        unique_neighbours = set()
        for p in L:
            for p1 in p.neigbours:
                for p2 in p1.neighbours:
                    if p2.MIS_neighbours() == 0:
                        unique_neighbours.add(p2)
        return list(unique_neighbours)

    def CalculateMISAfterEdgeRemove(self, p1, p2):
        p1.Calculate_MIS_neighbours()
        p2.Calculate_MIS_neighbours()
        if not p1.IsPartOfMis() and not p2.IsPartOfMis():
            pass
        elif p1.IsPartOfMis() and not p2.IsPartOfMis():
            if p2.MIS_neighbours() == 0 and p2.classification == CLASS_LOW:
                flag = False
                for neighbour in p2.mis_neighbours:
                    if neighbour.IsPartOfMis():
                        flag = True
                        break
                if not flag:
                    p2.SetPartOfMis(True)
        elif not p1.IsPartOfMis() and p2.IsPartOfMis():
            if p1.MIS_neighbours() == 0 and p1.classification == CLASS_LOW:
                flag = False
                for neighbour in p1.mis_neighbours:
                    if neighbour.IsPartOfMis():
                        flag = True
                        break
                if not flag:
                    p1.SetPartOfMis(True)

    def UpdateClassArrayTexts(self):
        self.lHighText.SetText("[" + ", ".join(self.highArr) + "]")
        self.lHighMedText.SetText("[" + ", ".join(self.highMedArr) + "]")
        self.lMedLowText.SetText("[" + ", ".join(self.medLowArr) + "]")
        self.lLowText.SetText("[" + ", ".join(self.lowArr) + "]")


    def ClassifyPoint(self, point):
        m = len(self.edges)
        degV = point.GetDegree()

        highValue = m ** 0.75
        medValue = m ** 0.5
        lowValue = m ** 0.25

        if(degV >= highValue):
            point.SetClassification(CLASS_HIGH)
        elif highValue > degV >= medValue:
            point.SetClassification(CLASS_HIGH_MED)
        elif medValue > degV >= lowValue:
            point.SetClassification(CLASS_MED_LOW)
        else:
            point.SetClassification(CLASS_LOW)
        self.RecollectClassArrays()

    def RecollectClassArrays(self):
        self.highArr = []
        self.highMedArr = []
        self.medLowArr = []
        self.lowArr = []

        for point in self.points:
            if point.classification == CLASS_HIGH:
                self.highArr.append(point.text)
            elif point.classification == CLASS_HIGH_MED:
                self.highMedArr.append(point.text)
            elif point.classification == CLASS_MED_LOW:
                self.medLowArr.append(point.text)
            else:
                self.lowArr.append(point.text)

        self.UpdateClassArrayTexts()

    def ChangeMis(self, event):
        for point in self.points:
            if point.IsInside(event.pos):
                if not point.IsPartOfMis():
                    if not any(x.IsPartOfMis() for x in point.neighbours) :
                        point.SetPartOfMis(True)
                else:
                    point.SetPartOfMis(False)

    def LoadDemo(self, event = None, context = None):
        self.ClearCanvas()
        p1 = self.CreatePoint(700, 140)
        p2 = self.CreatePoint(980, 345)
        p3 = self.CreatePoint(870, 675)
        p4 = self.CreatePoint(525, 675)
        p5 = self.CreatePoint(420, 340)

        p6 = self.CreatePoint(700, 260)
        p7 = self.CreatePoint(875, 385)
        p8 = self.CreatePoint(810, 590)
        p9 = self.CreatePoint(595, 590)
        p10 = self.CreatePoint(525, 390)

        self.CreateEdge(p1, p2)
        self.CreateEdge(p2, p3)
        self.CreateEdge(p3, p4)
        self.CreateEdge(p4, p5)
        self.CreateEdge(p5, p1)

        self.CreateEdge(p6, p8)
        self.CreateEdge(p7, p9)
        self.CreateEdge(p8, p10)
        self.CreateEdge(p9, p6)
        self.CreateEdge(p10, p7)

        self.CreateEdge(p1, p6)
        self.CreateEdge(p2, p7)
        self.CreateEdge(p3, p8)
        self.CreateEdge(p4, p9)
        self.CreateEdge(p5, p10)
































